### PR TITLE
New monster:zombie hollow[Ready]

### PR DESCRIPTION
**Pull request by Frost-wood**
Now that different groups of monsters can spawn based on time elapsed, monsters can be added that couldn't before because of newbie-destroying potential.

I humbly present to you the hollow zombie, a zombie that execute short-ranged attacks due to its lack of solid bones.  The idea behind this zombie is that the Blob has converted over time the inside of the zombie to something more resembling the blob.  The skin and shape is the only thing human that remains, with the skin being used more as a shell to protect XE037, from the outside world.

The blob has made the bones, muscles and flesh more flexible-as per thresh Slime mutations, which allows to arms and head to stretch out and reach the player from 1-3 tiles away.  Damage is a little weak, and I'm not sure whether I should increase it, hopefully people give me there thoughts on this.

I plan to add other monsters in separate PRs to prevent them from sitting in PR hell because a monster is unmergeable or doesn't fit in with the theme of the game.

### CODE REVIEW

File: data/json/monstergroups.json [hunk 1]
```diff
@@ -1359,7 +1359,8 @@
       { "monster" : "mon_zombie_master", "freq" : 10, "cost_multiplier" : 15 },
       { "monster" : "mon_beekeeper", "freq" : 5, "cost_multiplier" : 1 },
       { "monster" : "mon_zombie_technician", "freq" : 10, "cost_multiplier" : 6},
-      { "monster" : "mon_zombie_brute_shocker", "freq" : 10, "cost_multiplier" : 20 }
+      { "monster" : "mon_zombie_brute_shocker", "freq" : 10, "cost_multiplier" : 10 },
+	  { "monster" : "mon_zombie_hollow", "freq" : 2, "cost_multiplier" : 5 }
```
Comment (KA101) [inline data/json/monstergroups.json]: Tab check.  :-/

File: data/json/monsters.json [hunk 2]
```diff
@@ -2473,6 +2473,37 @@
...
+    "species":"ZOMBIE",
+    "default_faction":"zombie",
+    "symbol":"Z",
+    "color":"white",
+    "size":"MEDIUM",
+    "material":"flesh",
+    "diff":20,
+    "aggression":100,
+    "morale":100,
+    "speed":90,
```
Comment (KA101) [inline data/json/monsters.json]: OK, so these are much faster than a standard zed (75 speed) despite being non-rigid.  I'd expect 'em to move more slowly since they don't have bones, etc.

File: data/json/monsters.json [hunk 3]
```diff
@@ -2473,6 +2473,37 @@
...
+    "melee_dice":2,
+    "melee_dice_sides":6,
+    "melee_cut":4,
+    "dodge":4,
+    "armor_bash":8,
+    "armor_cut":8,
+    "vision_day":30,
+    "vision_night":3,
+    "item_chance":80,
+    "death_drops": "mon_zombie_soldier_death_drops",
```
Comment (KA101) [inline data/json/monsters.json]: Hmm.  Are these an upgraded soldier, and if so, would the armor not serve as an exoskeleton of sorts (and deserve mention in the blurb?)

File: src/monattack.cpp [hunk 4]
```diff
@@ -3560,6 +3560,84 @@ void mattack::bite(monster *z, int index)
...
+                foe->add_msg_player_or_npc( msg_type, _("The %s's head extends to bite you, but you dodge and the head sails past!"),
+                                                      _("The %s's head extends to bite <npcname>, but they dodge and the head sails past!"),
+                                            z->name().c_str() );
+            }
+            if( !uncanny ) {
+                foe->practice( "dodge", z->type->melee_skill * 2 );
+                foe->ma_ondodge_effects();
+            }
+        } else if( seen ) {
+            add_msg( _("The %s's head extends at %s, but misses and sails past!"), z->name().c_str(), target->disp_name().c_str() );
```
Comment (KA101) [inline src/monattack.cpp]: Seems like this "sails past" sort of miss might stun the zed for a turn or so, as it needs to retract the head?

**Comment by Rivet-the-Zombie** [comment]
These are a really neat idea!

File: src/monattack.cpp [hunk 5]
```diff
@@ -4217,10 +4295,81 @@ void mattack::suicide(monster *z, int index)
...
+                                                     _("<npcname> evades the stretched arm and it sails past <npcname>!"),
+                                                    z->name().c_str());
+                foe->practice( "dodge", z->type->melee_skill * 2 );
+                foe->ma_ondodge_effects();
+            } else {
+                add_msg( _("The %s evades the stretched arm and it sails past %s!"),
+                            target->disp_name().c_str());
+            }
+            //takes some time to retract the arm
+            z->moves -= 150;
```
Comment (KA101) [inline src/monattack.cpp]: OK, this works for the arm (which, presumably, doesn't have any sensory inputs).

File: src/monattack.cpp [hunk 6]
```diff
@@ -3560,6 +3560,84 @@ void mattack::bite(monster *z, int index)
...
+    dam = target->deal_damage( z, hit, damage_instance( DT_STAB, dam ) ).total_damage();
+
+    if( dam > 0 && foe != nullptr ) {
+        if( seen ) {
+            auto msg_type = foe == &g->u ? m_bad : m_info;
+            //~ 1$s is monster name, 2$s bodypart in accusative
+            foe->add_msg_player_or_npc( msg_type,
+                                        _("The %1$s's teeth sinks into your %2$s!"),
+                                        _("The %1$s's teeth sinks into <npcname>'s %2$s!"),
+                                        z->name().c_str(),
```
Comment (Marrim) [inline src/monattack.cpp]: Teeth are plural. So "sink", not "sinks".

**Comment by PropaneSoup** [comment]
I really like this, and based off of this I'm looking forward to any more you have lined up. :)
When it's butchered right now though, it drops tainted meat. Should probably be bits of blob if I'm understanding its description.

File: data/json/monsters.json (hunk 3 above)
Comment (Frost-wood) [inline data/json/monsters.json]: Whoops, wasn't supposed to have any clothing drops as I imagine the stretching and what-not would rip them.

**Comment by Frost-wood** [comment]
@PropaneSoup: I like the idea that the zombie is more blob than zombie so I changed the death function to blobsplit.

File: src/monattack.cpp [hunk 7]
```diff
@@ -3560,6 +3560,85 @@ void mattack::bite(monster *z, int index)
...
+void mattack::stretch_bite(monster *z, int index)
+{
+    if( !z->can_act() ) {
+        return;
+    }
+
+    // Let it be used on non-player creatures
+    // can be used at close range too!
+    Creature *target = z->attack_target();
+    if( target == nullptr || rl_dist( z->pos(), target->pos() ) > 3 ) {
```
Comment (Coolthulhu) [inline src/monattack.cpp]: Also needs a check for clear line here or else it'll bite through solid glass and iron bars.

File: src/monattack.cpp [hunk 8]
```diff
@@ -4217,10 +4296,81 @@ void mattack::suicide(monster *z, int index)
...
+        return;
+    }
+    Creature *target = z->attack_target();
+    if (target == nullptr || rl_dist(z->pos(), target->pos()) > 3 || !z->sees(*target)){
+        return;
+    }
+    int distance = rl_dist(z->pos(), target->pos());
+    player *foe = dynamic_cast< player* >( target );
+    bool seen = g->u.sees( *z );
+    if (distance > 2 && distance <= 3){
```
Comment (Coolthulhu) [inline src/monattack.cpp]: Why not just `distance == 3`?

File: data/json/monstergroups.json (hunk 1 above)
Comment (Frost-wood) [inline data/json/monstergroups.json]: tab seems to be fine in notepad++, don't know why git is displaying it differently

File: src/monattack.cpp (hunk 7 above)
Comment (KA101) [inline src/monattack.cpp]: Depending on whether it's armored, passing through bars might be OK (see: AMORPHOUS).  Agreed that glass, etc is not.

File: data/json/monstergroups.json [hunk 9]
```diff
@@ -1359,8 +1359,8 @@
       { "monster" : "mon_zombie_master", "freq" : 10, "cost_multiplier" : 15 },
       { "monster" : "mon_beekeeper", "freq" : 5, "cost_multiplier" : 1 },
       { "monster" : "mon_zombie_technician", "freq" : 10, "cost_multiplier" : 6},
-      { "monster" : "mon_zombie_brute_shocker", "freq" : 10, "cost_multiplier" : 20 }
-
+      { "monster" : "mon_zombie_brute_shocker", "freq" : 10, "cost_multiplier" : 10 },
```
Comment (KA101) [inline data/json/monstergroups.json]: **!**
*KA101 hurriedly makes a radio call*
I've spotted a stealth rebalance!  Send reinforcements!

(To be clear: nope, changing spawn levels that specifically should at least be specifically mentioned.  Ideally they'd be their own PR.  I'm not sure it's warranted as a result of adding these folks, so gonna ask you to revert it.)

File: src/monattack.cpp (hunk 4 above)
Comment (KA101) [inline src/monattack.cpp]: OK, thanks, this is taken care of at 3585-86 now.  (Note for myself/others, since Git didn't detect your fixing this line note.  Thanks.)

File: src/monattack.cpp [hunk 10]
```diff
@@ -4217,10 +4296,81 @@ void mattack::suicide(monster *z, int index)
...
     }
     z->die(z);
 }
+
+void mattack::stretch_attack(monster *z, int index){
+    if( !z->can_act() ) {
+        return;
+    }
+    Creature *target = z->attack_target();
+    if (target == nullptr || rl_dist(z->pos(), target->pos()) > 3 || !z->sees(*target)){
```
Comment (KA101) [inline src/monattack.cpp]: "Arms" should have no problem going through bars; chain fences, etc not so much.

**Comment by KA101** [comment]
Not yet ready for review, but looking worthwhile.  :-)

File: src/monattack.cpp (hunk 7 above)
Comment (Frost-wood) [inline src/monattack.cpp]: AMOPHOUS, thanks I was looking for something in the code that does that, and the closest thing I could find was the gun code.

**Comment by Frost-wood** [comment]
Tested, should work fine, the arms go through bars but are stopped by windows, chain-fence ,etc, and the head is the same, except it doesn't go through bars.
